docs/tutorials/event_driven.md

```diff
@@ -0,0 +1,298 @@
+        // This event is late; its window has already been triggered.
+    } else {
+        // Round up eventTime to the end of the window containing this event.
+        long endOfWindow = (eventTime - (eventTime % durationMsec) + durationMsec - 1);
+
+        // Schedule a callback for when the window has been completed.
+        timerService.registerEventTimeTimer(endOfWindow);
+
+        // Add this fare's tip to the running total for that window.
+        Float sum = sumOfTips.get(endOfWindow);
+        if (sum == null) {
+            sum = 0.0F;
+        }
+        sum += fare.tip;
+        sumOfTips.put(endOfWindow, sum);
+    }
+}
+{% endhighlight %}
+
+Things to consider:
+
+* What happens with late events? Events that are behind the watermark (i.e., late) are being
+  dropped. If you want to do something better than this, consider using a side output, which is
+  explained in the [next section]({{ site.baseurl }}{% link tutorials/event_driven.md
+  %}#side-outputs).
+
+* This example uses `MapState` where the keys are timestamps, and sets a `Timer` for that same
+  timestamp. This is a common pattern; it makes it easy and efficient to lookup relevant information
+  when the timer fires.
+
+#### The `onTimer()` method
+
+{% highlight java %}
+public void onTimer(
+        long timestamp, 
+        OnTimerContext context, 
+        Collector<Tuple3<Long, Long, Float>> out) throws Exception {
+
+    long driverId = context.getCurrentKey();
+    // Look up the result for the hour that just ended.
+    Float sumOfTips = this.sumOfTips.get(timestamp);
+
+    Tuple3 result = new Tuple3(driverId, timestamp, sumOfTips);
+    out.collect(result);
+    this.sumOfTips.remove(timestamp);
+}
+{% endhighlight %}
+
+Observations:
+
+* The `OnTimerContext context` passed in to `onTimer` can be used to determine the current key.
+
+* Our pseudo-windows are being triggered when the current watermark reaches the end of each hour, at
+  which point `onTimer` is called. This onTimer method removes the related entry from `sumOfTips`,
+  which has the effect of making it impossible to accommodate late events. This is the equivalent of
+  setting the allowedLateness to zero when working with Flink's time windows.
+
+### Performance Considerations
+
+Flink provides `MapState` and `ListState` types that are optimized for RocksDB. Where possible,
+these should be used instead of a `ValueState` object holding some sort of collection. The RocksDB
+state backend can append to `ListState` without going through ser/de, and for `MapState`, each
+key/value pair is a separate RocksDB object, so `MapState` can be efficiently accessed and updated.
+
+{% top %}
+
+## Side Outputs
+
+### Introduction
+
+There are several good reasons to want to have more than one output stream from a Flink operator, such as reporting:
+
+* exceptions
+* malformed events
+* late events
+* operational alerts, such as timed-out connections to external services
```

Comment:
All of these are some sort of errors, maybe also mention normal splits in application logic depending on what input type you have or so?